src/coap_tinydtls.c

```diff
@@ -562,11 +698,454 @@ coap_get_tls_library_version(void) {
+/* p   q   r   s   t   u   v   w   x   y   z                     */
+  41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+  64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64
+  };
+
+/* caller must free off returned coap_binary_t* */
+static coap_binary_t *
+pem_base64_decode (const uint8_t *data, size_t size)
+{
+  uint8_t *tbuf = coap_malloc(size);
+  size_t nbytesdecoded;
+  size_t i;
+  coap_binary_t *decoded;
+  uint8_t *ptr;
+  uint8_t *out;
+  size_t nb64bytes = 0;
+
+  for (i = 0; i < size; i++) {
+    switch (data[i]) {
+    case ' ':
+    case '\r':
+    case '\n':
+    case '\t':
+      break;
+    default:
+      if (b64_6[data[i]] == 64)
+        goto end;
+      tbuf[nb64bytes++] = data[i];
+      break;
+    }
+  }
+
+end:
+  nbytesdecoded = ((nb64bytes + 3) / 4) * 3;
+  decoded = coap_new_binary(nbytesdecoded + 1);
+  if (!decoded)
+    return NULL;
+
+  out = decoded->s;
+  ptr = tbuf;
+
+  while (nb64bytes > 4) {
+    *(out++) = b64_6[ptr[0]] << 2 | b64_6[ptr[1]] >> 4;
+    *(out++) = b64_6[ptr[1]] << 4 | b64_6[ptr[2]] >> 2;
+    *(out++) = b64_6[ptr[2]] << 6 | b64_6[ptr[3]];
+    ptr += 4;
+    nb64bytes -= 4;
+  }
+
+  /* Note: (nb64bytes == 1) is an error */
+  if (nb64bytes > 1) {
+    *(out++) = b64_6[ptr[0]] << 2 | b64_6[ptr[1]] >> 4;
+  }
+  if (nb64bytes > 2) {
+    *(out++) = b64_6[ptr[1]] << 4 | b64_6[ptr[2]] >> 2;
+  }
+  if (nb64bytes > 3) {
+    *(out++) = b64_6[ptr[2]] << 6 | b64_6[ptr[3]];
+  }
+
+  decoded->length = nbytesdecoded - ((4 - nb64bytes) & 3);
+  coap_free(tbuf);
+  return decoded;
+}
+
+static size_t
+asn1_len(const uint8_t **ptr)
+{
+  size_t len = 0;
+
+  if ((**ptr) & 0x80) {
+    size_t oct = (**ptr) & 0x7f;
+    (*ptr)++;
+    while (oct) {
+      len = (len << 8) + (**ptr);
+      (*ptr)++;
+      oct--;
+    }
+  }
+  else {
+    len = (**ptr) & 0x7f;
+    (*ptr)++;
+
+  }
+  return len;
+}
+
+static size_t
+asn1_tag_c(const uint8_t **ptr, int *constructed, int *class)
+{
+  size_t tag = 0;
+
+  *constructed = ((**ptr) & 0x20) ? 1 : 0;
+  *class = (**ptr) >> 6;
+  tag = (**ptr) & 0x1F;
+  (*ptr)++;
+  if (tag < 0x1F)
+    return tag;
+
+  tag = (**ptr) & 0x7F;
+  (*ptr)++;
+  return tag;
+}
+
+typedef coap_binary_t * (*asn1_callback)(const uint8_t *data, size_t size);
+
+static coap_binary_t *
+asn1_callback_privkey(const uint8_t *data, size_t size)
+{
+  coap_binary_t *priv_key;
+
+  /* Check if we have the private key (with a leading 0x00) */
```

Comment:
My private key is without this "leading 0x00".
Are you sure, it is required and could not be added?

Reply:
Proposal:

```
  /* skip leading 0x00 */
  if (size - 1 == DTLS_EC_KEY_SIZE) {
	--size;
	++data;
  }

  /* Check if we have the private key */
  if (size != DTLS_EC_KEY_SIZE)
    return NULL;

  priv_key = coap_new_binary(DTLS_EC_KEY_SIZE);
  if (!priv_key)
    return NULL;


  memcpy(priv_key->s, data, DTLS_EC_KEY_SIZE);
  return priv_key;

```

With that, I can read my 2. PEM. 

If the PUBLIC KEY section is not available, but the PRIVATE KEY section is tried, I guess this will works as well. At least it works, if the private data is copied into the public.

```
-----BEGIN EC PRIVATE KEY-----
MHcCAQEEIK7wwpfZinfUfzPDcChiU/FlLmvvUQhEMyYuwYx771OkoAoGCCqGSM49
AwEHoUQDQgAEyMjc2oOA4JKTj5EVD/9WYAS7aHI2mDNg/vPPQVDQh/7kLGc1ynTE
rZnPpc5Xk9bgs7SxKdkpjzWJwUjia9xs9A==
-----END EC PRIVATE KEY-----

-----BEGIN PUBLIC KEY-----
MHcCAQEEIK7wwpfZinfUfzPDcChiU/FlLmvvUQhEMyYuwYx771OkoAoGCCqGSM49
AwEHoUQDQgAEyMjc2oOA4JKTj5EVD/9WYAS7aHI2mDNg/vPPQVDQh/7kLGc1ynTE
rZnPpc5Xk9bgs7SxKdkpjzWJwUjia9xs9A==
-----END PUBLIC KEY-----
```

Reply:
My RPK that works does have the leading 0x00.  I know that the leading 0x00 is to be ignored but would like to know why it is (optionally) there.  

Your proposal looks good, and I will make the changes and test it out.

I separately am thinking about using the publickey embedded in the RPK (ignoring the entry in public_cert) to make things simpler to use.